### PR TITLE
add Dice to metrax (#64)

### PR DESCRIPTION
**Pull request by Mrigankkh**
### Summary [Issue #64 ]

- Added Dice metric as a part of Classification Metrics to metrax.
-  Support for Multidimensional vectors for predictions
-  I've followed the Keras implementation for Dice Loss to Compute the Dice coefficient. [In some implementations, epsilon is added in the numerator but since it it not added in Keras, I did not add it in this implementation either]
- Reused Classification tests and also added edge cases tests for All-ones, All_zeros and all_overlap. Tested against the Dice formula since Keras does not offer Dice as a metric.

### CODE REVIEW

File: src/metrax/classification_metrics.py [hunk 1]
```diff
@@ -577,3 +577,87 @@ def compute(self) -> jax.Array:
...
+          sum_pred=jnp.array(0.0, jnp.float32),
+          sum_true=jnp.array(0.0, jnp.float32),
+      )
+
+  @classmethod
+  def from_model_output(
+      cls,
+      predictions: jax.Array,
+      labels: jax.Array,
+      threshold: float = 0.5,
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:620]: seems like threshold isn't used in the keras implementation. their documentation is as follows:
```
        y_true: tensor of true targets.
        y_pred: tensor of predicted targets.
```

**Review by jshin1394 — COMMENTED** [review]
Thank you so much :)

File: src/metrax/classification_metrics.py [hunk 2]
```diff
@@ -577,3 +577,87 @@ def compute(self) -> jax.Array:
...
+  ) -> 'Dice':
+      """Updates the metric.
+
+      Args:
+          predictions: A floating point vector whose values are in the range [0,
+            1]. The shape should be (batch_size,).
+          labels: True value. The value is expected to be 0 or 1. The shape should
+            be (batch_size,).
+            weight of each sample. The shape should be (batch_size,).
```
[the remark this answers is not in the document]
Comment (Mrigankkh) [reply src/metrax/classification_metrics.py:629]: Will do! Sample_weights are a bit tricky since in keras, theyre implemented for a model-level loss computing keras pipeline.

 Looking to understand the entire mechanism a little better, I'll make these updates soon once i have a better understanding

[the remark this answers is not in the document]
Comment (Mrigankkh) [reply src/metrax/classification_metrics.py:629]: Hi Jiwon!
All of Metrax’s existing metrics treat sample_weights as element-level weights. Weights are multiplied into every entry before we form the numerator/denominator.
Keras Dice Loss applies weights after it has already computed a per-sample Dice score.

Should I stay consistent with Metrax’s other metrics (element-level weighting) or match Keras exactly (per-sample weighting)? Currently, I've implemented the first option, but for tests, when I tried using Keras Dice Loss to compute the Dice coeff, the tests fail due to this reason. I could compute expected results the same way for tests but I just wanted to clarify.

I’ll align the code and tests with whichever option you prefer. Thanks!

**Comment by Mrigankkh** [comment]
- Lint Checked
- Added sample_weights at element level and adjusted tests accordingly

File: src/metrax/classification_metrics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jshin1394) [reply src/metrax/classification_metrics.py:669]: Hi Mrigank,

Thanks for your investigation! To clarify your findings, could you please explain what you mean by "element-level weights" and "per-sample mean"?

Specifically, if "per-sample" refers to a single (y_pred , y_true) pair from the input batch, could DICE then inherit from Metrax's base.Average and compute the DICE score for each individual (y_pred, y_true) pair within the batch?

[the remark this answers is not in the document]
Comment (Mrigankkh) [reply src/metrax/classification_metrics.py:669]: By element level I mean broadcasting sample_weights so that every (y_pred, y_true) is individually weighed before computing DICE and per-sample [the way keras implements it] is calculating unweighted DICE and then applying sample_weights. 

I've implemented element level i.e. broadcasting sample_weights for every y_pred/y_true pair. This is in line with the other metrics that also compute metrics in element-level. 

We may be able to use base.Average if we switch to the per-sample method where we first calculated unweighted DICE and then broadcast sample_weights. This is closer to what Keras does but most other metrics in Metrax broadcast sample_weights before any computation.

 Keras does a bunch of under the hood stuff when  base keras.losses.Loss.__call__  is called (ex: computing masks which as per my understanding uses some internal attributes) which is then reused while computing DICE loss; I was not sure how I could mirror keras's implementation exactly which is why I took the other approach after tracing how other metrics in Metrax handle sample_weights

[the remark this answers is not in the document]
Comment (jshin1394) [reply src/metrax/classification_metrics.py:669]: I see. Thank you for such detailed explanation. Let's leave out sample_weights as part of current implementation as I do not want to make any assumptions about our users needs at the moment. We will support DICE without sample_weights for now, and we will add support once users request for it with specific use cases(element-wise, sample-wise, etc).

File: src/metrax/classification_metrics_test.py [hunk 3]
```diff
@@ -78,7 +81,14 @@ def test_aucroc_empty(self):
     self.assertEqual(m.false_positives, jnp.array(0, jnp.float32))
     self.assertEqual(m.false_negatives, jnp.array(0, jnp.float32))
     self.assertEqual(m.num_thresholds, 0)
-
```
Comment (jshin1394) [inline src/metrax/classification_metrics_test.py:81]: nit: add extra line between test cases

File: src/metrax/classification_metrics.py [hunk 4]
```diff
@@ -577,3 +577,82 @@ def compute(self) -> jax.Array:
     )
     # Threshold goes from 0 to 1, so trapezoid is negative.
     return jnp.trapezoid(tp_rate, fp_rate) * -1
+
+
+@flax.struct.dataclass
+class Dice(clu_metrics.Metric):
```
Comment (jshin1394) [inline src/metrax/classification_metrics.py:583]: let's put dice under image_metrics for now. Note how we have __init__ such that users can just refer to Metrax.Dice either way.

**Review by jshin1394 — APPROVED** [review]
Almost there! thank you :)

File: src/metrax/classification_metrics_test.py (hunk 3 above)
Comment (Mrigankkh) [reply src/metrax/classification_metrics_test.py:81]: There is only 1 line between all test cases. Do you want me to add extra lines for all test cases?
